### ISSUE TITLE
reading from console hangs build

### ISSUE THREAD

**Issue by veyga** [propensive/fury-old#1394]
Build process hangs if input is coming from console. 

object Hey extends App {
  println("enter your name")
  val name = io.StdIn.readLine
  println(s"hello $name")
}

fury build run --output linear
Starting compilation of module console/consolemod
Successfully compiled module console/consolemod
enter your name
andrew

aasdfasdf
.......

All output formats result in this effect. Ctrl-c exits the process, but its left running (displayed in jps -l).

**Comment by propensive** [comment]
This is an important issue. Can you check what happens if you run Fury in "standalone" mode, using the Fury launcher instead of an installed version?, i.e. use `./fury build run --output linear`. This will invoke Fury using its `main` method in a new JVM, instead of using Nailgun, which may be contributing to the problems.

But I suspect that's not the problem. We start a new `java` process to run the application, and I don't think we join its `STDIN` to the System `STDIN` anywhere.

**Comment by veyga** [comment]
same issue. will try to fix

**Comment by veyga** [comment]
Since setting the default stream can't be done through java command line options, I think this requires edits to guillotine. The process builder sets stdout and stderr, but not stdin. The input stream from nailgun probably needs to be passed all the way to this call to async in guillotine.

**Comment by propensive** [comment]
Yep, I'm not surprised. We can make changes to Guillotine if you want to have a play around there, too.

**Comment by propensive** [comment]
And the changes on the Fury side shouldn't be too huge.

**Comment by veyga** [comment]
how is fury initially resolving the outside dependencies (guillotine, magnolia, etc)?
They don't show up as imported layers, nor as imported repos.

**Comment by propensive** [comment]
It imports the `platform` layer, which imports most of these other layers (which may themselves import other layers too). You can make changes to guillotine like so:
```
fury layer select -l /platform/guillotine
fury repo fork -r guillotine -f ../guillotine
```
then start editing the sources in `../guillotine`.

**Comment by veyga** [comment]
This is a surprisingly wicked problem; not sure how this is possible with the current setup. 
I'm experimenting with both redirecting subprocess pipe as well as reading from nailgun in/writing to the subprocess out.  I'm trying to run java/scala console apps with various input sources (scala.io.StdIn, Scanner, inputstreamreaders),  but they aren't working properly, as the nailgun input is either not being passed to subprocess stdin, or is immediately sending a null terminator to the application (which is fine for any app that doesn't read from console).

Unfortunately, nailgun does not allow one to set standard in manually. Normally you can redirect subprocess or inherit from the parent process, but nailgun makes this virtually impossible since it utilizes an entity called ThreadLocalInputStream, which throws exceptions all over the place if you try to mess with stdin/stdout/stderr. There might be a workaround, but I haven't gotten it working yet. I've tried every possble combo between passing the input stream around, passing around functions to bind the subprocess to the nailgun input stream, piping, etc. I also tried setting up NIO channels, but nailgun uses synchronous IO, so its not simple to convert. 

It would seem simple to set this up, but alas. Since you can't directly set input streams given nailgun, I'm afraid I don't know how this will work. I would think writing to the process's stdin from nailgun input stream would work, but the control flow seems locked or out of order. The current "flow" expects a stream of compileevents, and doesn't really account for stopping to read from input, which is likely the reason for an immediate null terminator being sent.
